JCB-based obs+bias staging, Jedi class updates, and marine B-matrix refactoring

DavidNew-NOAA opened the pull request:
# Description

This PR is a companion to GDASApp PR [#1312](https://github.com/NOAA-EMC/GDASApp/pull/1312) (merged) and JCB-GDAS PR [#31](https://github.com/NOAA-EMC/jcb-gdas/pull/31) (merged).

This PR does three things:

1. It changes the observation and bias staging for the atmospheric analysis tasks to use JCB templates instead of reading the full JEDI input configuration dictionary in order to construct a list of files to stage. This is cleaner and places fewer constraints on how to initialize the analysis.
2. The ```Jedi``` constructor now takes as input a dictionary that is essentially subset of the ```task_config``` dictionary. This makes the code clearer and less opaque and makes debugging easier. Each dictionary is constructed from a YAML file with configuration parameters for each JEDI application that is run.
3. All JEDI applications and their input YAMLs are now initialized in the initialize job of the ```AtmAnalysis``` and ```AtmEnsAnalysis```. Before, in the ```atmensanl*``` jobs for example, the LETKF solver was initialized in the ```atmensanlinit```cjob, but the LETKF solver and FV3 increment converter were both initialized and executed in the ```atmensanlobs``` and ```atmensanlfv3inc``` jobs respectively. This makes more sense in terms of resource allocation.

Addendum:

I'm now rolling in the refactoring of the marine B-matrix task into this PR. That makes it also a companion of GDASApp PR [#1346](https://github.com/NOAA-EMC/GDASApp/pull/1346) and JCB-GDAS PR [#36](https://github.com/NOAA-EMC/jcb-gdas/pull/36).

These new changes introduce the ```Jedi``` class and JCB into the marine B-matrix job.

Partially resolves GDASApp issue [#1296](https://github.com/NOAA-EMC/GDASApp/issues/1296)

# Type of change
- [ ] Bug fix (fixes something broken)
- [X] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? YES
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO
  - [ ] EMC verif-global <!-- NOAA-EMC/EMC_verif-global#1234 -->
  - [X] GDAS [#1346](https://github.com/NOAA-EMC/GDASApp/pull/1346)
  - [ ] GFS-utils <!-- NOAA-EMC/gfs-utils#1234 -->
  - [ ] GSI <!-- NOAA-EMC/GSI#1234 -->
  - [ ] GSI-monitor <!-- NOAA-EMC/GSI-Monitor#1234 -->
  - [ ] GSI-utils <!-- NOAA-EMC/GSI-Utils#1234 -->
  - [ ] UFS-utils <!-- ufs-community/UFS_UTILS#1234 -->
  - [ ] UFS-weather-model <!-- ufs-community/ufs-weather-model#1234 -->
  - [ ] wxflow <!-- NOAA-EMC/wxflow#1234 -->

# How has this been tested?
C96C48_ufs_hybatmDA CI runs successfully
GDASApp jjob tests pass successfully

# Checklist
- [ ] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have documented my code, including function, input, and output descriptions
- [X] My changes generate no new warnings
- [X] New and existing tests pass with my changes
- [X] This change is covered by an existing CI test or a new one has been added
- [ ] I have made corresponding changes to the system documentation if necessary

RussTreadon-NOAA committed:
replace separate JEDI radiance bias correction files with tarball (#2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
rename JEDI radiance bias correction tarball file to be more self-describing (#2862)

RussTreadon-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
update gdas.cd hash to include recent commits to GDASApp develop (#2862)

RussTreadon-NOAA committed:
update pygfs jedi class and atmospheric tasks to extract JEDI radiance bias correction files from tarball (#2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
move radiance bias correction staging to jedi class (#2868)

RussTreadon-NOAA committed:
initial attempt to generalize processing of variational bias correction files using jedi class (#2862)

RussTreadon-NOAA committed:
move VarBcDir inside initialize_analysis method (#2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
update sorc/gdas.cd hash to bring in radiance tarball ctest changes (#2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
add extract_tar method to jedi class, use extract_tar in atm and atmens analysis scripts (#2862)

RussTreadon-NOAA committed:
correct typo in atmens_analysis.py (2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
improve error handling for jedi class tarfile.extractall (#2862)

RussTreadon-NOAA committed:
remove unnecessary close following tarfile add (#2862)

RussTreadon-NOAA committed:
use endswith to clean up scripting (#2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
refactor jedi class method extract_tar (#2862)

RussTreadon-NOAA committed:
update sorc/gdas.cd to current head of GDASApp develop (#2862)

RussTreadon-NOAA committed:
make jedi method extract_tar static, clean up comments, correct typo (#2862)

RussTreadon-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/radbcor

RussTreadon-NOAA committed:
correctly prefix tlapse radiance bias correction files for cycling (#2862)

RussTreadon-NOAA committed:
extend duration of C96C48_ufs_hybatmDA CI by one cycle to 2024022406 (#2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

DavidNew-NOAA committed:
Initial commit

DavidNew-NOAA committed:
Merge branch 'feature/radbcor' into feature/jcb-obsbias

RussTreadon-NOAA committed:
correct ORION.env typo, adjust JEDI ORION job configurations (#2862)

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/radbcor

DavidNew-NOAA committed:
Merge branch 'feature/radbcor' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Update

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Initialize all JEDI applications on the initialize jobs

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
pynorms #2

DavidNew-NOAA committed:
pynorms #3

DavidNew-NOAA committed:
Fix some python bugs

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Fixing bugs

DavidNew-NOAA committed:
Bug fix

DavidNew-NOAA committed:
Remove redundant code for tarball extraction, update comments, and update GDAS hash

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
pynorms #2

DavidNew-NOAA commented:
I'm leaving this PR as draft until the companion GDASApp PR is merged, but it is ready for review otherwise. cc @aerorahul @WalterKolczynski-NOAA @RussTreadon-NOAA

CoryMartin-NOAA commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -22,249 +20,260 @@ class Jedi:
     @logit(logger)
-    def extract_tar(tar_file: str) -> None:
-        """Extract files from a tarball
+    def extract_tar_from_fh_dict(fh_dict) -> None:
```
picking nits, but `fh` to me means "forecast hour" at first glance, perhaps `file_dict`?

CoryMartin-NOAA commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -22,249 +20,260 @@ class Jedi:
+                if os.path.isfile(tar_file):
+                    logger.info(f"Extract files from {tar_file}")
+                    extract_tar(tar_file)
+                else:
```
can we raise a `FileNotFound` exception?

CoryMartin-NOAA commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -22,249 +20,260 @@ class Jedi:
+
+
+@logit(logger)
+def extract_tar(tar_file: str) -> None:
```
@kevindougherty-noaa FYI

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Address Cory's comments

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Slight change for readability

DavidNew-NOAA committed:
Add key checking to JEDI class constructor

DavidNew-NOAA committed:
Update

DavidNew-NOAA committed:
Update

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
Initial commit

DavidNew-NOAA committed:
Merge branch 'develop' into feature/marine-refactor

DavidNew-NOAA committed:
Debug

DavidNew-NOAA committed:
Merge branch 'feature/jcb-obsbias' into feature/marine-refactor

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Saving progress

DavidNew-NOAA committed:
Saving progress

DavidNew-NOAA committed:
Update GDAS hash to develop

DavidNew-NOAA commented:
@aerorahul @WalterKolczynski-NOAA @RussTreadon-NOAA With all companion PRs now merged, I'm marking this ready for review.

DavidNew-NOAA committed:
Clean up exception handling

DavidNew-NOAA committed:
Save progress

DavidNew-NOAA committed:
Merge branch 'develop' into feature/marine-refactor

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Merge branch 'develop' into feature/marine-refactor

DavidNew-NOAA committed:
Bugfixes

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
Merge branch 'develop' into feature/marine-refactor

DavidNew-NOAA committed:
Remove comment

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Add comment blocks to methods

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
Make jcb rendering method save copies of jcb_config and task_ocnfig

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
Merge branch 'feature/jcb-obsbias' into feature/marine-refactor

DavidNew-NOAA commented:
@guillaumevernieres and @AndrewEichmann-NOAA , FYI, I'm rolling the marine B-matrix refactoring into this PR since that refactoring depends on the changes made to the ```Jedi``` class in this PR.

The relevant files changed for the B-matrix refactoring are:
```parm/config/gfs/config.marineanl```
```parm/config/gfs/config.marinebmat```
```scripts/exglobal_marinebmat.py```
```ush/python/pygfs/task/marine_analysis.py``` (very minor JCB-related change)
```ush/python/pygfs/task/marine_bmat.py```

guillaumevernieres reviewed (dismissed):
Looks good. Thanks @DavidNew-NOAA .

DavidNew-NOAA committed:
Fix to make sure bkg_list.yaml isn't loaded directly in marine JCB base

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update gdas hash

DavidNew-NOAA committed:
Adding back accidentally deleted line

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update GDAS hash

AndrewEichmann-NOAA reviewed (dismissed):
No complaints

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

CoryMartin-NOAA commented on ush/python/pygfs/task/atm_analysis.py:
```diff
@@ -73,46 +70,42 @@ def __init__(self, config: Dict[str, Any], yaml_name: Optional[str] = None):
-        None
-        """
+        # Create dictionary of JEDI objects
+        self.jedi = AttrDict()
```
I think we might want a better name for this object, but don't have a good suggestion. `self.jedi_tasks` maybe? just some way to make it clearer that it is a dictionary of Jedi class objects

DavidNew-NOAA replied:
@CoryMartin-NOAA Good call, will change

WalterKolczynski-NOAA commented on ush/python/pygfs/task/atm_analysis.py:
```diff
@@ -73,46 +70,42 @@ def __init__(self, config: Dict[str, Any], yaml_name: Optional[str] = None):
-        logger.debug(f"Writing JEDI YAML config to: {self.jedi.yaml}")
-        save_as_yaml(self.jedi.config, self.jedi.yaml)
+        # atmanlvar
+        self.jedi['atmanlvar'] = Jedi(AttrDict(
```
I think this whole thing is a sign this class is trying to do too many independent things. The class should only be instantiated for increment or for variational (one executable, one algo yaml, etc.), not try to cover both as is being done here. If the differences are just in these settings (as seems to be the case), this can be done without a new class by just passing in the relevant values. If there is coding that needs to be different, it can with child classes that inherit from `AtmAnalysis`.

Right now, I don't think the two are even created in the same job, so this could be as easy as using the same envvar to specify the executable and the algo yaml in the separate config files for each job instead of separate names for each type. i.e. call them `JEDIEXE` and `JCB_ALGO_YAML`, and just specify them differently in the config file. Since the configs are only sourced in their respective jobs, there would be no name collision (and each could still use a different `@_@` variable, if you )

This should allow you to revert a lot of these changes that are adding messiness to the method calls and keep things nice and straightforward.

DavidNew-NOAA replied:
@WalterKolczynski-NOAA What you're seeing is indeed one Jedi object per JEDI executable. AtmAnalysis.jedi, for example, is a dictionary of Jedi objects, each key pointing to one object for one executable. This structure, along with initializing each JEDI object with a dictionary of required parameters (executable path, JCB algo, etc), will eventually allow us to put all these parameters in one big, clean YAML file to configure all the JEDI executable for a set of jobs like the atm or marine analysis. Eventually we can make a static method in the Jedi class that constructs this dictionary of Jedi objects from that YAML.

The other reason this structure and for getting rid of JEDIEXE, for example, as a generic environment variable and replacing with JEDIEXE_* for each executable, is this allows us to initialize all JEDI executables in the init job. The way things were before in AtmEnsAnalysis, for example, we would initialize the atmensanlobs executable in the atmensanlinit job, but the rest of the JEDI executable would have to be initialized in their respective run jobs. For example, the atmensanlfv3inc executable had to be initialized in the atmensanlfv3inc job where that executable is actually run. I think we'd like to keep all initialization in the atmensanlinit job. Having one JEDIEXE and one JCB_ALGO per config file prevents us from doing this.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/atmens_analysis.py:
```diff
@@ -73,46 +71,71 @@ def __init__(self, config: Dict[str, Any], yaml_name: Optional[str] = None):
-        ----------
-        None
+        # atmensanlobs
+        self.jedi['atmensanlobs'] = Jedi(AttrDict(
```
Same as above. One object = one configuration.

WalterKolczynski-NOAA commented on ush/python/pygfs/task/marine_bmat.py:
```diff
@@ -38,18 +57,114 @@ def __init__(self, config):
+        self.jedi = AttrDict()
+
+        # gridgen
+        self.jedi['gridgen'] = Jedi(AttrDict(
```
Same here, though since I think these are called in the same job, you'll need to pass the correct configuration in when the each object is created.

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update gdas hash

DavidNew-NOAA committed:
Move Jedi object initialization dictionaries into YAML files for AtmAnalysis and AtmEnsAnalysis

DavidNew-NOAA committed:
Move marine bmat JEDI config to YAML file

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA commented:
Some new commits. gdas jjob tests and marine CI all pass.

@CoryMartin-NOAA The ```Jedi``` object dictionaries are now all named ```AtmAnalysis.jedi_dict```, ```AtmEnsAnalysis.jedi_dict```, ```MarineBMat.jedi_dict``` to indicate that they are dictionaries of ```Jedi``` objects, thereby addressing your comment.

@WalterKolczynski-NOAA I've now moved all the parameters for the ```Jedi``` initialization dictionaries into three YAML files, ```parm/gdas/atmanl_jedi_config.yaml.j2```, ```parm/gdas/atmensanl_jedi_config.yaml.j2```, and ```parm/gdas/soca_bmat_jedi_config.yaml.j2```. A static method in the ```Jedi``` class, ```Jedi.get_jedi_dict```, is now used to convert those YAML files each into a dictionary of ```Jedi``` objects. This cleans up a lot of the mess at the beginning of the ```AtmAnalysis```, ```AtmEnsAnalysis```, and  ```MarineBMat``` classes, initializing the ```Jedi``` objects. This further makes things more transparent and ultimately moves us closer to unified YAML-based configuration of jobs.

This also eliminates the problem of namespace collision of the```JEDIEXE```, ```JCB_ALGO```, etc environment variables (which now no longer exist) so that we can initialize all JEDI executables in the init jobs for each set of jobs (atm analysis, atmens analysis, marine analysis, etc). I hope this addresses the comment you made last night.

Hopefully, we can get @aerorahul 's eyes on this PR early next week and move this it forward, so we can get the aero, snow, and remaining components of marine analysis refactored according to this framework.

cc @guillaumevernieres @AndrewEichmann-NOAA

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Minor change to retrigger GW CI

WalterKolczynski-NOAA commented:
> Some new commits. gdas jjob tests and marine CI all pass.
> 
> @CoryMartin-NOAA The `Jedi` object dictionaries are now all named `AtmAnalysis.jedi_dict`, `AtmEnsAnalysis.jedi_dict`, `MarineBMat.jedi_dict` to indicate that they are dictionaries of `Jedi` objects, thereby addressing your comment.
> 
> @WalterKolczynski-NOAA I've now moved all the parameters for the `Jedi` initialization dictionaries into three YAML files, `parm/gdas/atmanl_jedi_config.yaml.j2`, `parm/gdas/atmensanl_jedi_config.yaml.j2`, and `parm/gdas/soca_bmat_jedi_config.yaml.j2`. A static method in the `Jedi` class, `Jedi.get_jedi_dict`, is now used to convert those YAML files each into a dictionary of `Jedi` objects. This cleans up a lot of the mess at the beginning of the `AtmAnalysis`, `AtmEnsAnalysis`, and `MarineBMat` classes, initializing the `Jedi` objects. This further makes things more transparent and ultimately moves us closer to unified YAML-based configuration of jobs.
> 
> This also eliminates the problem of namespace collision of the`JEDIEXE`, `JCB_ALGO`, etc environment variables (which now no longer exist) so that we can initialize all JEDI executables in the init jobs for each set of jobs (atm analysis, atmens analysis, marine analysis, etc). I hope this addresses the comment you made last night.
> 
> Hopefully, we can get @aerorahul 's eyes on this PR early next week and move this it forward, so we can get the aero, snow, and remaining components of marine analysis refactored according to this framework.
> 
> cc @guillaumevernieres @AndrewEichmann-NOAA

This isn't what I was looking for. Each `AtmAnalysis` should only have one `JEDI` object inside, not a list, and that object should not be referenced outside of the `AtmAnalysis` class. `initialize()` and `finalize()` would be no-argument calls on the instantiation, not static methods. They should all look like this (possibly spread across different jobs):
```
atm_anl = AtmAnalysis(config)
atm_anl.initialize()
atm_anl.execute()
atm_anl.finalize()
```

Anything needed for those three steps are passed in as part of `config` and stored in the object for when the methods get called. No additional information should be needed once it has been configured at instantiation. If you need both variational and increment objects (I don't see a need in these, but if you do), they would be different objects:
```
atm_var = AtmAnalysis(config_var)
atm_increment = AtmAnalysis(config_increment)
atm_var.initialize()
atm_increment.initialize()
...
```

If any of the three methods need to do substantially different things, we should instead create child classes of `AtmAnalysis`, but keep the same philosophy:
```
class AtmAnalysisVar(AtmAnalysis):
    def __init__(self, config: Dict[str, Any]):
        super.__init__(config)
        # Additional init code unique to var

    def initialize(self):
        super.initialize()
        # Additional initialize code unique to var
    ....
```
```
atm_var = AtmAnalysisVar(config_var)
atm_increment = AtmAnalysisInc(config_increment)
...
```

DavidNew-NOAA committed:
Address Walter's comments

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA commented:
Thanks, @WalterKolczynski-NOAA  , for clarifying your suggestions.

My latest commits overnight wrap the execute method of the ```Jedi``` class into execute methods for the ```AtmAnalysis```, ```AtmEnsAnalysis```, and ```MarineBMat``` classes. Thus you have something like ```AtmAnl.execute_var()```, ```AtmAnl.execute_fv3inc()```, for example, in the atm analysis ex scripts. I agree with you that it's better to simplify the ex scripts and not pass arguments to methods of those classes.

However, one purpose of this PR is in fact to move away from having one ```Jedi``` object per ```AtmAnalysis```, ```AtmEnsAnalysis```, ```MarineBMat``` class, where one passes ```JEDIEXE```, ```JEDI_ALGO```, etc environment variables through the constructors. In practice, that philosophy makes the workings of those classes quite opaque and difficult to understand. It's not always clear which JEDI application you're working with when you manipulate the ```Jedi``` object in those classes. It's also really restrictive and forces one to code in awkward ways.

Besides, the ```MarineBMat``` object executes something like 7 JEDI applications. It's too impractical to have 7 different ```MarineBMat``` objects going at once.

CoryMartin-NOAA commented:
may I suggest that it might be a good idea for @aerorahul @WalterKolczynski-NOAA and @DavidNew-NOAA to have a short meeting next week to confirm the design philosophy going forward?

DavidNew-NOAA commented:
@CoryMartin-NOAA Yeah, that's a good idea

DavidNew-NOAA committed:
Update gdas hash

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Update gdas hash and revert accidentally changes gfs_utils hash

DavidNew-NOAA committed:
Small, mostly asthetic changes

DavidNew-NOAA committed:
Address requested changes

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA committed:
Fix bug Russ found

DavidNew-NOAA committed:
Forgot a line

DavidNew-NOAA commented:
@aerorahul and @WalterKolczynski-NOAA I made updates to this PR based on your feedback yesterday. This PR is ready for review. Thanks

aerorahul commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -4,145 +4,177 @@
+
     @logit(logger, name="Jedi")
-    def __init__(self, task_config: AttrDict, yaml_name: Optional[str] = None) -> None:
+    def __init__(self, config) -> None:
```
```suggestion
    def __init__(self, config: Dict[str, Any]) -> None:
```

aerorahul commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -4,145 +4,177 @@
-        - set an empty AttrDict for the JEDI config
-        - set the default directory for J2-YAML templates
+        - create the jedi_config AttrDict and extend it with additional required entries
+        - save a coy of jedi_config
```
```suggestion
        - save a copy of jedi_config
```

aerorahul commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -4,145 +4,177 @@
+        """
+
+        # Fill JCB base YAML template and build JCB config dictionary
+        if self.jedi_config.jcb_base_yaml:
```
The pythonic way to check if `self.jedi_config` has an attribute `jcb_base_yaml` is to do:
```suggestion
        if hasattr(self.jedi_config, 'jcb_base_yaml'):
```

DavidNew-NOAA replied:
@aerorahul It's actually checking for ```None```. The ```Jedi``` class constructor ensures that all the required keys, including ```jcb_base_yaml``` are present. See here https://github.com/DavidNew-NOAA/global-workflow/blob/227718caf71c188a0075007bab9745b1f3468ba8/ush/python/pygfs/jedi/jedi.py#L47

aerorahul commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -4,145 +4,177 @@
+            raise KeyError(f"FATAL ERROR: JEDI configuration dictionary must contain jcb_base_yaml.")
+
+        # Add JCB algorithm YAML, if it exists, to JCB config dictionary
+        if self.jedi_config.jcb_algo_yaml:
```
same as above.

DavidNew-NOAA replied:
@aerorahul Same as my comment above

DavidNew-NOAA replied:
@aerorahul An exception shouldn't be raised here if ```jcb_algo_yaml``` is set to ```None```. JCB algorithm YAMLs are not required. It's only required that the key, ```algorithm```, is present somewhere in the jcb_config dict, and it's only sometimes specified in the JCB algorithm YAML.

aerorahul commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -152,119 +184,150 @@ def link_exe(task_config: AttrDict) -> None:
+        if expected_keys:
+            for jedi_dict_key in expected_keys:
+                if jedi_dict_key not in jedi_dict:
+                    raise Exception(f"FATAL ERROR: {jedi_dict_key} not present {jedi_config_yaml}")
```
log this with `logger.execption` or `logger.error` and then raise the exception.

aerorahul commented on ush/python/pygfs/jedi/jedi.py:
```diff
@@ -152,119 +184,150 @@ def link_exe(task_config: AttrDict) -> None:
+        if tarfile.is_tarfile(tar_file):
+            logger.error(f"FATAL ERROR: {tar_file} could not be read")
+            raise tarfile.ReadError(f"FATAL ERROR: unable to read {tar_file}")
+        else:
+            logger.info()
```
What is the benefit/use of this `else` block?

DavidNew-NOAA replied:
@RussTreadon-NOAA Do you know what was supposed to go in this else block? I think this code dates back to [#2875](https://github.com/NOAA-EMC/global-workflow/pull/2875)

DavidNew-NOAA replied:
I was avail to get the jist of what was supposed to go here and filled it in in my latest commit

aerorahul reviewed (commented):
@DavidNew-NOAA 
A partial review of files I could review this AM.  I will continue reviewing after my AM meeting.

DavidNew-NOAA committed:
Update ush/python/pygfs/jedi/jedi.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

DavidNew-NOAA committed:
Update ush/python/pygfs/jedi/jedi.py

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

DavidNew-NOAA committed:
Log errors before raising exceptions

DavidNew-NOAA committed:
Make validation of JEDI class configuration dictionary more robust

DavidNew-NOAA committed:
pynorms+

DavidNew-NOAA committed:
Fill in missing code for tar archive extraction

DavidNew-NOAA committed:
pynorms

DavidNew-NOAA commented:
@aerorahul OK, changes address your comments and also made validation of the input dictionary for the ```Jedi``` class constructor more robust

DavidNew-NOAA committed:
Update GDAS hash

DavidNew-NOAA committed:
Update gdas hash

DavidNew-NOAA committed:
Update gdas hash

DavidNew-NOAA committed:
Forgot to remove old execute methods

DavidNew-NOAA committed:
Get rid of execute_vtscales method

DavidNew-NOAA committed:
pynorms

aerorahul reviewed (dismissed):
looks ok to me.

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA commented:
@WalterKolczynski-NOAA The issue with the CI failures is the update of the GDASApp hash, since the old hash points to some patches for its submodules. We should hold off on CI until [#1362](https://github.com/NOAA-EMC/GDASApp/issues/1362) is resolved. @RussTreadon-NOAA is using this branch as his baseline and is making rapid progress. Sorry, I should have let you known ealier.

guillaumevernieres commented:
> @WalterKolczynski-NOAA The issue with the CI failures is the update of the GDASApp hash, since the old hash points to some patches for its submodules. We should hold off on CI until [#1362](https://github.com/NOAA-EMC/GDASApp/issues/1362) is resolved. @RussTreadon-NOAA is using this branch as his baseline and is making rapid progress. Sorry, I should have let you known ealier.

@DavidNew-NOAA , any reason why we need to have the new variable name as part of this pr? If not I'd suggest to revert to a working gdas.cd # so we can move the other pr's through the pipeline.

DavidNew-NOAA commented:
@guillaumevernieres This PR depends on a more recent GDASApp hash due to GDASApp PR [#1312](https://github.com/NOAA-EMC/GDASApp/pull/1312) (merged), but I think we could throw together a quick GDASApp PR that would at least fix these CI failures. 

The ```enkfgdas_esnowrecen``` failure in ```C96C48_hybatmaerosnowDA``` is due to a bug in the FMS2 IO interface in FV3-JEDI. I made a quick bugfix PR [#1304](https://github.com/JCSDA-internal/fv3-jedi/pull/1304). It's approved, but I think Francois was just waiting for JCSDA CI to pass, which it did.

The failure in ```gdas_atmanlinit``` in ```C96C48_ufs_hybatmDA``` is due to lack of bias corrections for ```abi_g16```, but that's commented out in the latest hash by GDASApp PR [#1363](https://github.com/NOAA-EMC/GDASApp/pull/1363).

The failure in ```gdas_aeroanlvar``` and ```gfs_aeroanlvar``` in ```C96C48_hybatmaerosnowDA``` is due to https://github.com/NOAA-EMC/GDASApp/blob/develop/parm/io/fv3jedi_fieldmetadata_restart.yaml having ```DELP``` rather than ```delp``` for its io name. That's the only one I'm not sure that if we change it, we'll break other stuff. Russ' https://github.com/NOAA-EMC/GDASApp/tree/feature/resume_nightly branch changes it, but it also updates all the JEDI hashes associated with the variable naming sprint. Any thoughts on that particular detail @RussTreadon-NOAA  ?

RussTreadon-NOAA commented:
I've been logging progress in getting g-w DA CI to pass with `DavidNew-NOAA:feature/jcb-obsbias` using GDASApp `feature/resume_nightly` in GDASApp issue [#1362](https://github.com/NOAA-EMC/GDASApp/issues/1362).  As noted in [#1362](https://github.com/NOAA-EMC/GDASApp/issues/1362) I am currently working through issues with `gdas_marineanlvar`.

guillaumevernieres commented:
> @guillaumevernieres This PR depends on a more recent GDASApp hash due to GDASApp PR [#1312](https://github.com/NOAA-EMC/GDASApp/pull/1312) (merged), but I think we could throw together a quick GDASApp PR that would at least fix these CI failures.
> 
> The `enkfgdas_esnowrecen` failure in `C96C48_hybatmaerosnowDA` is due to a bug in the FMS2 IO interface in FV3-JEDI. I made a quick bugfix PR [#1304](https://github.com/JCSDA-internal/fv3-jedi/pull/1304). It's approved, but I think Francois was just waiting for JCSDA CI to pass, which it did.
> 
> The failure in `gdas_atmanlinit` in `C96C48_ufs_hybatmDA` is due to lack of bias corrections for `abi_g16`, but that's commented out in the latest hash by GDASApp PR [#1363](https://github.com/NOAA-EMC/GDASApp/pull/1363).
> 
> The failure in `gdas_aeroanlvar` and `gfs_aeroanlvar` in `C96C48_hybatmaerosnowDA` is due to https://github.com/NOAA-EMC/GDASApp/blob/develop/parm/io/fv3jedi_fieldmetadata_restart.yaml having `DELP` rather than `delp` for its io name. That's the only one I'm not sure that if we change it, we'll break other stuff. Russ' https://github.com/NOAA-EMC/GDASApp/tree/feature/resume_nightly branch changes it, but it also updates all the JEDI hashes associated with the variable naming sprint. Any thoughts on that particular detail @RussTreadon-NOAA ?

Ha ... yes ... Thanks for the reminders. I also just remembered that @RussTreadon-NOAA had to cook a saber/vader # in a fork ... Never mind my comment above!

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

RussTreadon-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA committed:
Try to fix marineanlletkf

DavidNew-NOAA committed:
Follow example of ORION.env for HERCULUS.env marineanlletkf

RussTreadon-NOAA commented:
FYI @DavidNew-NOAA

I am manually running g-w DA CI on Hera & Hercules using g-w `feature/jcb-obsbias` at 42904ba0 with `sorc/gdas.cd` populated with GDASApp `feature/resume_nightly` at [`716dcdb`](https://github.com/NOAA-EMC/GDASApp/commit/716dcdb0d3181e05256f73ea23c84283211216a3).

`test_gdasapp` had 64 out of 64 tests pass on both machines.

`test_gdasapp` is currently running on Orion.  Pending a 64/64 result I'll launch g-w DA CI on Orion.

DavidNew-NOAA commented:
@RussTreadon-NOAA Awesome! Getting closer...

RussTreadon-NOAA commented:
64/64 `test_gdasapp` passed on Orion.  g-w DA CI launched on Orion.   g-w DA CI on Hercules may not work.  The stage jobs are still PENDING in the batch queue.  Hercules is swamped with jobs.

We can't build GDASApp on Cactus following the system upgrade  - see GDASApp issue [#1376](https://github.com/NOAA-EMC/GDASApp/issues/1376) and g-w issue #3100.

RussTreadon-NOAA commented:
**g-w DA CI on Hera**

All jobs in all four streams successfully ran to completion

```
rocotostat /scratch1/NCEPDEV/stmp2/Russ.Treadon/EXPDIR/prgsi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202112201800        Done    Nov 15 2024 00:37:19    Nov 15 2024 00:55:16
202112210000        Done    Nov 15 2024 00:37:19    Nov 15 2024 03:20:29
202112210600        Done    Nov 15 2024 00:37:19    Nov 15 2024 03:00:26

rocotostat /scratch1/NCEPDEV/stmp2/Russ.Treadon/EXPDIR/prjedi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202402231800        Done    Nov 15 2024 00:35:18    Nov 15 2024 00:50:17
202402240000        Done    Nov 15 2024 00:35:18    Nov 15 2024 03:40:33
202402240600        Done    Nov 15 2024 00:35:18    Nov 15 2024 03:40:33

rocotostat /scratch1/NCEPDEV/stmp2/Russ.Treadon/EXPDIR/praero_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202112201200        Done    Nov 15 2024 00:35:20    Nov 15 2024 00:55:20
202112201800        Done    Nov 15 2024 00:35:20    Nov 15 2024 03:15:28
202112210000        Done    Nov 15 2024 00:35:20    Nov 15 2024 03:00:31

rocotostat /scratch1/NCEPDEV/stmp2/Russ.Treadon/EXPDIR/prwcda_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202103241200        Done    Nov 15 2024 00:35:22    Nov 15 2024 00:50:24
202103241800        Done    Nov 15 2024 00:35:22    Nov 15 2024 01:55:24
```

`PSLOT` maps to the following CI configurations:
- `prgsi`: _C96C48_hybatmDA_
- `prjedi`: _C96C48_ufs_hybatmDA_
- `praero`: _C96C48_hybatmaerosnowDA_
- `prwcda`: _C48mx500_3DVarAOWCDA_

RussTreadon-NOAA commented:
g-w DA CI for _C96C48_ufs_hybatmDA_ on Orion uncovered a bug in `parm/config/gfs/config.resources.ORION`.  

The `prjedi_pr2992` gfs_atmanlvar and gdas_atmanlvar jobs failed with OOM kills.  This was due to an incorrect job configuration in the xml.  The xml had 
```
<nodes>3:ppn=40:tpp=1</nodes>
```
This is not correct.  The configuration should be
```
<nodes>8:ppn=12:tpp=1</nodes>
```

A check of `parm/config/gfs/config.resources.ORION` showed that it contains.
```
   "atmanlvar")
    # Run on 8 nodes for memory requirement
    export tasks_per_node=8
```
This is incorrect.  The `$RUN` suffix needs to be appended to `tasks_per_node`.  That is,
```
  "atmanlvar")
    # Run fewer task per node for memory
    export tasks_per_node_gdas=8
    export tasks_per_node_gfs=8
```
With this change the correct job configuration is written to the xml file.

The corrected `config.resources.ORION` is on Orion in `/work2/noaa/da/rtreadon/git/global-workflow/pr2992/parm/config/gfs/`

 @DavidNew-NOAA , would you please commit the corrected `config.resources.ORION` to `DavidNew-NOAA:feature/jcb-obsbias`

DavidNew-NOAA committed:
Fixed bug in orion resource allocation

DavidNew-NOAA committed:
Merge branch 'develop' into feature/jcb-obsbias

DavidNew-NOAA commented:
@RussTreadon-NOAA Done! Good catch

RussTreadon-NOAA commented:
**Note**

With the update to JEDI hashes in `sorc/gdas.cd` comes the need to update JEDI fix files in `fix/gdas/fv3jedi`.  g-w issue #3104 has been opened to request addition of updated `fv3jedi` fix files.   Once the updated fix files are in place `gdas_fv3jedi_ver` in `versions/fix.ver` will need to be updated accordingly.

RussTreadon-NOAA commented:
**g-w DA CI on Orion**

All jobs in all four streams successfully ran to completion
```
rocotostat /work2/noaa/stmp/rtreadon/EXPDIR/prgsi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202112201800        Done    Nov 15 2024 02:40:34    Nov 15 2024 03:05:05
202112210000        Done    Nov 15 2024 02:40:34    Nov 15 2024 05:40:05
202112210600        Done    Nov 15 2024 02:40:34    Nov 15 2024 05:35:05

rocotostat /work2/noaa/stmp/rtreadon/EXPDIR/prjedi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202402231800        Done    Nov 15 2024 02:40:37    Nov 15 2024 03:05:07
202402240000        Done    Nov 15 2024 02:40:37    Nov 15 2024 14:05:07
202402240600        Done    Nov 15 2024 02:40:37    Nov 15 2024 14:51:52

rocotostat /work2/noaa/stmp/rtreadon/EXPDIR/praero_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202112201200        Done    Nov 15 2024 02:40:39    Nov 15 2024 03:10:10
202112201800        Done    Nov 15 2024 02:40:39    Nov 15 2024 05:50:09
202112210000        Done    Nov 15 2024 02:40:39    Nov 15 2024 05:40:09

rocotostat /work2/noaa/stmp/rtreadon/EXPDIR/prwcda_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202103241200        Done    Nov 15 2024 02:40:42    Nov 15 2024 03:05:11
202103241800        Done    Nov 15 2024 02:40:42    Nov 15 2024 04:20:12
```

RussTreadon-NOAA commented:
**g-w DA CI on Hercules**

All jobs in all four streams successfully ran to completion

```
rocotostat /work/noaa/stmp/rtreadon/EXPDIR/prgsi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202112201800        Done    Nov 15 2024 00:49:03    Nov 15 2024 04:25:04
202112210000        Done    Nov 15 2024 00:49:03    Nov 15 2024 13:05:02
202112210600        Done    Nov 15 2024 00:49:03    Nov 15 2024 14:30:04

rocotostat /work/noaa/stmp/rtreadon/EXPDIR/prjedi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202402231800        Done    Nov 15 2024 00:49:04    Nov 15 2024 04:25:05
202402240000        Done    Nov 15 2024 00:49:04    Nov 15 2024 14:50:04
202402240600        Done    Nov 15 2024 00:49:04    Nov 15 2024 16:15:04

rocotostat /work/noaa/stmp/rtreadon/EXPDIR/praero_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202112201200        Done    Nov 15 2024 00:49:05    Nov 15 2024 10:55:05
202112201800        Done    Nov 15 2024 00:49:05    Nov 15 2024 14:30:06
202112210000        Done    Nov 15 2024 00:49:05    Nov 15 2024 15:18:27

rocotostat /work/noaa/stmp/rtreadon/EXPDIR/prwcda_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202103241200        Done    Nov 15 2024 00:49:06    Nov 15 2024 04:30:06
202103241800        Done    Nov 15 2024 00:49:06    Nov 15 2024 12:55:06
```

RussTreadon-NOAA commented:
**g-w DA CI on WCOSS2 (Dogwood)**

Install `DavidNew-NOAA:feature/jcb-obsbias` at 7e6207f on Dogwood with `sorc/gdas.cd` populated with `feature/resume_nightly` at [`716dcdb`](https://github.com/NOAA-EMC/GDASApp/commit/716dcdb0d3181e05256f73ea23c84283211216a3)

All jobs in the following two g-w DA CI configurations successfully ran to completion
- prgsi: _C96C48_hybatmDA_
- prjedi: _C96C48_ufs_hybatmDA_

```
rocotostat /lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prgsi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201800        Done    Nov 15 2024 14:11:30    Nov 15 2024 14:40:10
202112210000        Done    Nov 15 2024 14:11:30    Nov 15 2024 16:35:08
202112210600        Done    Nov 15 2024 14:11:30    Nov 15 2024 16:20:14
 
rocotostat /lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prjedi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202402231800        Done    Nov 15 2024 14:11:31    Nov 15 2024 14:40:13
202402240000        Done    Nov 15 2024 14:11:31    Nov 15 2024 17:01:18
202402240600        Done    Nov 15 2024 14:11:31    Nov 15 2024 16:51:33
```

As expected jobs in the following g-w DA CI configurations failed
- praero: _C96C48_hybatmaerosnowDA_
- prwcda: _C48mx500_3DVarAOWCDA_

Specifically the following jobs failed
```
rocotostat /lfs/h2/emc/ptmp/russ.treadon/EXPDIR/praero_pr2992
202112201200        gdas_aeroanlgenb                   206170398                DEAD                 -29         2        1850.0
202112201800            gdas_snowanl                   206167952                DEAD                   1         2          69.0
202112201800             gfs_snowanl                   206167953                DEAD                   1         2          68.0
 
rocotostat /lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prwcda_pr2992
202103241800         gdas_marinebmat                   206169971                DEAD                 -29         2        1852.0
```

Recompile GDASApp with the test `spack-stack/1.6.0` installation and rerun the above DEAD jobs.  Each failed job in successfully ran to completion.  Resume both praero and prwcda CI runs.   

prwcda gdas_marineanlvar failed because `APRUN_MARINEANLVAR` was not defined.  Examine `env/WCOSS2.env`.  Jobs `marineanlvar` and `marineanlchkpt` were not present.  Replaced `ocnanalrun` and `ocnanalchkpt` with their marine counterparts.  Failed marineanlvar sucessfully ran to completion.

@davidnew, we should update `env/WCOSS2.env`.  Dogwood `/lfs/h2/emc/da/noscrub/russ.treadon/git/global-workflow/pr2992/env/WCOSS2.env` contains the following changes.
```
@@ -107,17 +107,15 @@ elif [[ "${step}" = "marinebmat" ]]; then
     export APRUNCFP="${launcher} -n \$ncmd --multi-prog"
     export APRUN_MARINEBMAT="${APRUN_default}"
 
-elif [[ "${step}" = "ocnanalrun" ]]; then
+elif [[ "${step}" = "marineanlvar" ]]; then
 
     export APRUNCFP="${launcher} -n \$ncmd --multi-prog"
+    export APRUN_MARINEANLVAR="${APRUN_default}"
 
-    export APRUN_OCNANAL="${APRUN_default}"
-
-elif [[ "${step}" = "ocnanalchkpt" ]]; then
+elif [[ "${step}" = "marineanlchkpt" ]]; then
 
     export APRUNCFP="${launcher} -n \$ncmd --multi-prog"
-
-    export APRUN_OCNANAL="${APRUN_default}"
+    export APRUN_MARINEANLCHKPT="${APRUN_default}"
 
 elif [[ "${step}" = "ocnanalecen" ]]; then
```


Eventually all jobs in praero and prwcda successfully ran to completion
```
rocotostat /lfs/h2/emc/ptmp/russ.treadon/EXPDIR/praero_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200        Done    Nov 15 2024 14:11:33    Nov 15 2024 17:23:45
202112201800        Done    Nov 15 2024 14:11:33    Nov 15 2024 19:11:59
202112210000        Done    Nov 15 2024 14:11:33    Nov 15 2024 19:04:32
 
rocotostat /lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prwcda_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241200        Done    Nov 15 2024 14:11:34    Nov 15 2024 14:40:19
202103241800        Done    Nov 15 2024 14:11:34    Nov 15 2024 18:00:23
```

DavidNew-NOAA committed:
Update APRUN commands in WCOSS2 env

DavidNew-NOAA commented:
@RussTreadon-NOAA Done

DavidNew-NOAA commented:
@RussTreadon-NOAA Sorry, wrong commit, one second

DavidNew-NOAA committed:
Update

DavidNew-NOAA committed:
update (#2)

RussTreadon-NOAA commented:
This PR depends on GDASApp PR [#1355](https://github.com/NOAA-EMC/GDASApp/pull/1355).  Once PR [#1355](https://github.com/NOAA-EMC/GDASApp/pull/1355) is approved and merged into GDASApp `develop`, the `sorc/gdas.cd` hash in this PR need to be updated to point at the new GDASApp `develop`.

DavidNew-NOAA committed:
Update gdas hash

RussTreadon-NOAA reviewed (dismissed):
Approve.  Changes tested via g-w DA CI.

DavidNew-NOAA commented:
@WalterKolczynski-NOAA @aerorahul With GDASApp PR [#1355](https://github.com/NOAA-EMC/GDASApp/pull/1355) merged into develop, this PR is tested and ready for final review.

aerorahul commented:
Does the group wish to let this run via CI on Hercules/Hera/Orion or bypass that?

RussTreadon-NOAA commented:
Once g-w issue #3104 is closed, we can update `versions/fix.ver` as shown below
```
--- a/versions/fix.ver
+++ b/versions/fix.ver
@@ -8,7 +8,7 @@ export cice_ver=20240416
 export cpl_ver=20230526
 export datm_ver=20220805
 export gdas_crtm_ver=20220805
-export gdas_fv3jedi_ver=20241022
+export gdas_fv3jedi_ver=20241115
 export gdas_soca_ver=20240802
 export gdas_gsibec_ver=20240416
 export gdas_obs_ver=20240213
```

The updated JEDI hashes in `sorc/gdas.cd` need updates to some fv3jedi fix files.

RussTreadon-NOAA commented:
> Does the group wish to let this run via CI on Hercules/Hera/Orion or bypass that?

Since we updated hashes and have one more change to commit (`versions/fix.ver`), I'd like to run Hera g-w DA CI to ensure all is well on at least one platform.  I'm fine with running label-based CI on Hercules, Hera, and Orion if the group would like to do so.

CoryMartin-NOAA commented:
My vote at 3:15pm Friday is to update the `fix.ver` before COB, add the labels, and let it run over the weekend. If all pass, merge at 9am Monday.

RussTreadon-NOAA committed:
Update gdas_fv3jedi_ver

RussTreadon-NOAA reviewed (approved):
Approve after `fix.ver` update

aerorahul committed:
Merge branch 'develop' into feature/jcb-obsbias

RussTreadon-NOAA commented:
**Hercules C96C48_hybatmDA failure**

Examine log files
```
/work2/noaa/global/CI/HERCULES/2992/RUNTESTS/COMROOT/C96C48_hybatmDA_32534bb9/logs/2021122106/enkfgdas_fcst_mem001.log
/work2/noaa/global/CI/HERCULES/2992/RUNTESTS/COMROOT/C96C48_hybatmDA_32534bb9/logs/2021122106/enkfgdas_fcst_mem002.log
```

Both forecast jobs died near the start of the model integration with nan messages
```
0: PASS: fcstRUN phase 1, n_atmsteps =                9 time is        12.491381
 0:  in atmos_model update, fhzero=   6.00000000000000      fhour=   3.333333
 0:   0.0000000E+00
 0: PASS: fcstRUN phase 2, n_atmsteps =                9 time is         3.004835
 4:
 4: FATAL from PE     4: NaN in input field of mpp_reproducing_sum(_2d), this indicates numerical instability
 4:
 4: Image              PC                Routine            Line        Source
 4: ufs_model.x        00000000056109A9  mpp_mod_mp_mpp_er          72  mpp_util_mpi.inc
 4: ufs_model.x        00000000058DBC7E  mpp_efp_mod_mp_mp         195  mpp_efp.F90
 4: ufs_model.x        0000000005824A7C  mpp_domains_mod_m         143  mpp_global_sum.fh
 4: ufs_model.x        0000000002F8D680  fv_grid_utils_mod        3077  fv_grid_utils.F90
 4: ufs_model.x        0000000002FC6949  fv_mapz_mod_mp_la         794  fv_mapz.F90
 4: libiomp5.so        0000150A777F6F13  __kmp_invoke_micr     Unknown  Unknown
 4: libiomp5.so        0000150A77766664  __kmp_fork_call       Unknown  Unknown
 4: libiomp5.so        0000150A7771FD23  __kmpc_fork_call      Unknown  Unknown
 4: ufs_model.x        0000000002FC1E3B  fv_mapz_mod_mp_la         683  fv_mapz.F90
 4: ufs_model.x        0000000002F69276  fv_dynamics_mod_m         771  fv_dynamics.F90
 4: ufs_model.x        0000000002EB982B  atmosphere_mod_mp         688  atmosphere.F90
 4: ufs_model.x        0000000002DA8B58  atmos_model_mod_m         879  atmos_model.F90
 4: ufs_model.x        0000000002C1F7E7  module_fcst_grid_        1335  module_fcst_grid_comp.F90
 4: ufs_model.x        0000000000ADA2E4  Unknown               Unknown  Unknown
 4: ufs_model.x        0000000000ADE25F  Unknown               Unknown  Unknown
 4: ufs_model.x        000000000114E75A  Unknown               Unknown  Unknown
 4: ufs_model.x        00000000012E3358  Unknown               Unknown  Unknown
 4: ufs_model.x        0000000000ADB73A  Unknown               Unknown  Unknown
 4: ufs_model.x        00000000009ED880  Unknown               Unknown  Unknown
 4: ufs_model.x        0000000000DC5641  Unknown               Unknown  Unknown
 4: ufs_model.x        0000000002C0D8D1  fv3atm_cap_mod_mp        1085  fv3_cap.F90
 4: ufs_model.x        0000000002C0D33B  fv3atm_cap_mod_mp        1034  fv3_cap.F90
```

This is an odd failure.  C96C48_hybatmDA is GSI-based atmospheric DA.  This PR does not alter either the forecast model or GSI-based DA.

Install DavidNew-NOAA:feature/jcb-obsbias at 32534bb9 on Hercules in `/work/noaa/da/rtreadon/git/global-workflow/pr2992_final`.  Set up and run C96C48_hybatmDA CI.  C96C48_hybatmDA successfully ran all jobs.
```
(gdasapp) hercules-login-2:/work/noaa/stmp/rtreadon/EXPDIR/prgsi_pr2992_final$ rocotostat -d prgsi_pr2992_final.db -w prgsi_pr2992_final.xml -c all -s
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202112201800        Done    Nov 16 2024 11:58:06    Nov 16 2024 13:00:04
202112210000        Done    Nov 16 2024 11:58:06    Nov 16 2024 19:04:02
202112210600        Done    Nov 16 2024 11:58:06    Nov 16 2024 18:44:03
```

Examine output from `/work2/noaa/global/CI/HERCULES/2992/RUNTESTS/COMROOT/C96C48_hybatmDA_32534bb9/`.  The `enkf.x` analysis increments for 20211221 06Z are nonphysical
```
 0:  time level            1
 0:  --------------
 0: ens. mean anal. increment min/max  u   -12204.3056641        7737.84765625
 0: ens. mean anal. increment min/max  v   -11370.2402344        8705.36914062
 0: ens. mean anal. increment min/max  tv   -2873.78125000        4715.19384766
 0: ens. mean anal. increment min/max  q  -0.978207349777       0.497514724731
 0: ens. mean anal. increment min/max  oz  -0.121405441314E-02   0.924816820771E-03
 0: ens. mean anal. increment min/max  ps   -1131.41528320        623.390930176
 0:  time level            2
 0:  --------------
 0: ens. mean anal. increment min/max  u   -10770.1357422        8994.37695312
 0: ens. mean anal. increment min/max  v   -9961.04296875        12255.5898438
 0: ens. mean anal. increment min/max  tv   -2885.21777344        6381.37304688
 0: ens. mean anal. increment min/max  q   -1.11263489723       0.543511331081
 0: ens. mean anal. increment min/max  oz  -0.126553326845E-02   0.859888270497E-03
 0: ens. mean anal. increment min/max  ps   -947.946472168        614.040649414
 0:  time level            3
 0:  --------------
 0: ens. mean anal. increment min/max  u   -9989.00585938        8761.36914062
 0: ens. mean anal. increment min/max  v   -7961.80371094        11510.4707031
 0: ens. mean anal. increment min/max  tv   -3515.27587891        5955.34912109
 0: ens. mean anal. increment min/max  q   -1.03635215759       0.862098753452
 0: ens. mean anal. increment min/max  oz  -0.934842391871E-03   0.881655723788E-03
 0: ens. mean anal. increment min/max  ps   -765.077331543        606.946655273
```
Compare this with the successful rerun
```
0:  time level            1
 0:  --------------
 0: ens. mean anal. increment min/max  u   -53.3018913269        49.1635665894
 0: ens. mean anal. increment min/max  v   -52.7409667969        49.9739608765
 0: ens. mean anal. increment min/max  tv   -16.9043731689        9.42595672607
 0: ens. mean anal. increment min/max  q  -0.238635810092E-02   0.335678178817E-02
 0: ens. mean anal. increment min/max  oz  -0.139211806527E-05   0.174592651092E-05
 0: ens. mean anal. increment min/max  ps   -2.33502197266        4.34594726562
 0:  time level            2
 0:  --------------
 0: ens. mean anal. increment min/max  u   -34.5535125732        29.5668640137
 0: ens. mean anal. increment min/max  v   -38.0788955688        39.3475875854
 0: ens. mean anal. increment min/max  tv   -14.1956558228        13.5581207275
 0: ens. mean anal. increment min/max  q  -0.306937331334E-02   0.292858737521E-02
 0: ens. mean anal. increment min/max  oz  -0.134557058118E-05   0.189299998965E-05
 0: ens. mean anal. increment min/max  ps   -2.89700317383        5.60110473633
 0:  time level            3
 0:  --------------
 0: ens. mean anal. increment min/max  u   -33.2891349792        65.5625305176
 0: ens. mean anal. increment min/max  v   -40.9432754517        40.9033317566
 0: ens. mean anal. increment min/max  tv   -18.5473480225        11.8623199463
 0: ens. mean anal. increment min/max  q  -0.335096800700E-02   0.276062171906E-02
 0: ens. mean anal. increment min/max  oz  -0.137578490467E-05   0.206520826396E-05
 0: ens. mean anal. increment min/max  ps   -3.58447265625        7.21514892578
```
Comparison of the enkfstat files from the two runs show no differences apart from timing statistics, different order of printout, and ensemble mean analysis increments.  

I have not seen this error in my many runs of C96C48_hybatmDA  on Hercules. A rerun of C96C48_hybatmDA via automated CI is likely to succeed.  This, however, sidesteps the above failure.  Why / how nonphysical increments were generated in this case remains unresolved.

RussTreadon-NOAA commented:
@WalterKolczynski-NOAA - What's the path forward from here?  We would like to get this PR merged into `develop` sooner rather than later.

- I do not see a way to only run C96C48_hybatmDA via git CI labels.  It's a waste of time and resources to rerun all configurations on Hercules.  Do we accept the explanation above, manually rerun C96C48_hybatmDA, or modify g-w CI to only run a single configuration?

- WCOSS2 dev (Dogwood) goes down tomorrow (12/19).  Do we launch g-w CI on Dogwood today?  My understanding is that WCOSS2 g-w CI is manual, not automatic.  Is this correct?

- Do we need to run CI on Orion?

RussTreadon-NOAA commented:
@aerorahul, @WalterKolczynski-NOAA , and @KateFriedman-NOAA 

I manually launched g-w DA CI on WCOSS2 (Dogwood).  We know C96C48_hybatmaerosnowDA and C48mx500_3DVarAOWCDA will fail.  GDASApp must be built with spack-stack on WCOSS2 in order for select jobs in these configurations to successfully run to complete.  GSI and JEDI based atmospheric DA, C96C48_hybatmDA and C96C48_ufs_hybatmDA, should pass.

The following g-w CI configuration do **NOT** contain a wcoss2 _skip_
- C48_ATM.yaml
- C48_S2SW.yaml
- C96_atm3DVar_extended.yaml

Do these three configurations also need to be run on Dogwood?

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA - What's the path forward from here? We would like to get this PR merged into `develop` sooner rather than later.
> 
> * I do not see a way to only run C96C48_hybatmDA via git CI labels.  It's a waste of time and resources to rerun all configurations on Hercules.  Do we accept the explanation above, manually rerun C96C48_hybatmDA, or modify g-w CI to only run a single configuration?
> * WCOSS2 dev (Dogwood) goes down tomorrow (12/19).  Do we launch g-w CI on Dogwood today?  My understanding is that WCOSS2 g-w CI is manual, not automatic.  Is this correct?
> * Do we need to run CI on Orion?

Unfortunately, the CI does not (yet) have the capability to select individual tests. We have to rerun them all on a platform.

Orion is not needed. Orion's support level has been downgraded and most of the things that would cause a failure on Orion will also cause a failure on Hercules anyway.

> @aerorahul, @WalterKolczynski-NOAA , and @KateFriedman-NOAA
> 
> I manually launched g-w DA CI on WCOSS2 (Dogwood). We know C96C48_hybatmaerosnowDA and C48mx500_3DVarAOWCDA will fail. GDASApp must be built with spack-stack on WCOSS2 in order for select jobs in these configurations to successfully run to complete. GSI and JEDI based atmospheric DA, C96C48_hybatmDA and C96C48_ufs_hybatmDA, should pass.
> 
> The following g-w CI configuration do **NOT** contain a wcoss2 _skip_
> 
> * C48_ATM.yaml
> * C48_S2SW.yaml
> * C96_atm3DVar_extended.yaml
> 
> Do these three configurations also need to be run on Dogwood?

Any tests that aren't being skipped should be run. It looks like the upgrade it postponed, so we should be able to get these in. I will run them overnight so hopefully they are complete by morning.

WalterKolczynski-NOAA commented:
CI Tests set up to run in /lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_2992/RUNTESTS on WCOSS

RussTreadon-NOAA commented:
**WCOSS2 g-w CI**
Install `DavidNew-NOAA:feature/jcb-obsbias` at 32534bb on Dogwood.  Run the following g-w CI configurations
-  _C96C48_hybatmDA_ - prgsi_pr2992
-  _C96C48_ufs_hybatmDA_ - prjedi_pr2992
-  _C96C48_hybatmaerosnowDA_ - praero_pr2992
-  _C48mx500_3DVarAOWCDA_ - prwcda_pr2992
-  _C48_ATM_ - pratm_pr2992
-  _C48_S2SW_ - prswsw_pr2992
-  _C96_atm3DVar_extended_ - pratm3dvar_pr2992

with the following results
```
/lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prgsi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201800        Done    Nov 18 2024 15:05:08    Nov 18 2024 15:25:14
202112210000        Done    Nov 18 2024 15:05:08    Nov 18 2024 20:55:05
202112210600        Done    Nov 18 2024 15:05:08    Nov 18 2024 20:35:11
 
/lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prjedi_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202402231800        Done    Nov 18 2024 15:05:19    Nov 18 2024 15:25:35
202402240000        Done    Nov 18 2024 15:05:19    Nov 18 2024 21:05:15
202402240600        Done    Nov 18 2024 15:05:19    Nov 18 2024 21:05:15
 
/lfs/h2/emc/ptmp/russ.treadon/EXPDIR/praero_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201200      Active    Nov 18 2024 15:05:30             -          
202112201800      Active    Nov 18 2024 15:05:30             -          
202112210000      Active    Nov 18 2024 15:05:30             -          
 
/lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prwcda_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103241200        Done    Nov 18 2024 15:05:36    Nov 18 2024 15:25:48
202103241800      Active    Nov 18 2024 15:05:36             -          
 
/lfs/h2/emc/ptmp/russ.treadon/EXPDIR/pratm_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    Nov 18 2024 16:35:30    Nov 18 2024 20:50:15
202103231800        Done    Nov 18 2024 16:35:30    Nov 18 2024 21:00:16
 
/lfs/h2/emc/ptmp/russ.treadon/EXPDIR/pratm3dvar_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202112201800        Done    Nov 18 2024 16:35:33    Nov 18 2024 17:20:21
202112210000        Done    Nov 18 2024 16:35:33    Nov 19 2024 00:00:22
202112210600        Done    Nov 18 2024 16:35:33    Nov 19 2024 01:40:22
202112211200        Done    Nov 18 2024 17:25:29    Nov 19 2024 04:05:16
202112211800        Done    Nov 19 2024 00:05:15    Nov 19 2024 07:35:15
 
/lfs/h2/emc/ptmp/russ.treadon/EXPDIR/prs2sw_pr2992
   CYCLE         STATE           ACTIVATED              DEACTIVATED     
202103231200        Done    Nov 18 2024 16:35:35    Nov 18 2024 20:50:26
202103231800        Done    Nov 18 2024 16:35:35    Nov 18 2024 21:00:29
``` 

All jobs in all configurations except _C96C48_hybatmaerosnowDA_ and _C48mx500_3DVarAOWCDA_ successfully ran to completion.  

Note that _C96C48_hybatmaerosnowDA_ and _C48mx500_3DVarAOWCDA_ are set to be **_skipped_** on WCOSS2.  If GDASApp is rebuilt and run with `spack-stack/1.6.0` these two configurations successfully run all jobs to completion.

The following configurations in `ci/cases/pr` list `wcoss2` under the `skip_ci_on_hosts:` section
```
./C96C48_hybatmaerosnowDA.yaml:  - wcoss2
./C96C48_ufs_hybatmDA.yaml:  - wcoss2
./C96_S2SWA_gefs_replay_ics.yaml:  - wcoss2
./C48mx500_3DVarAOWCDA.yaml:  - wcoss2
./C48mx500_hybAOWCDA.yaml:  - wcoss2
./C96_atm3DVar.yaml:  - wcoss2
./C48_S2SWA_gefs.yaml:  - wcoss2
```
These ci configurations, apart from the three configurations which include JEDI based DA, were not run on Dogwood.

Based on the above g-w CI results we can change the CI-Wcoss2-Running label to CI-Wcoss2-Passed.

RussTreadon-NOAA commented:
Acceptable (explainable) g-w CI results have been obtained on Hera, Hercules, and WCOSS2 (Dogwood).  

@WalterKolczynski-NOAA , @aerorahul , @KateFriedman-NOAA :  Are any more tests required before this PR can be merged into `develop`?

RussTreadon-NOAA commented:
Thank you @WalterKolczynski-NOAA
